presto-main/src/main/java/com/facebook/presto/operator/OperatorContext.java

```diff
@@ -409,24 +413,41 @@ public synchronized long requestMemoryRevoking()
     {
         boolean alreadyRequested = isMemoryRevokingRequested();
         if (!alreadyRequested && revocableMemoryReservation > 0) {
-            memoryRevokingRequestedFuture.set(null);
+            memoryRevokingRequested.set(true);
+            runListener();
             return revocableMemoryReservation;
         }
         return 0;
     }
 
     public synchronized void resetMemoryRevokingRequested()
     {
-        SettableFuture<?> currentFuture = memoryRevokingRequestedFuture;
-        if (!currentFuture.isDone()) {
-            return;
+        if (memoryRevokingRequested.compareAndSet(true, false)) {
+            memoryRevokingRequestedListener.set(null);
         }
-        memoryRevokingRequestedFuture = SettableFuture.create();
     }
 
-    public synchronized SettableFuture<?> getMemoryRevokingRequestedFuture()
+    public synchronized void setMemoryRevokingRequestedListener(Runnable listener)
     {
-        return memoryRevokingRequestedFuture;
+        verify(memoryRevokingRequestedListener.get() == null, "A listener has already been set");
+        memoryRevokingRequestedListener.set(listener);
+        // if memory revoking is requested immediately run the listener
+        if (memoryRevokingRequested.get()) {
+            runListener();
```

Comment:
I don't like foreign call under a lock. Fortunately this is here the last instruction in this synchronized section (_currently_), but generally it can lead to subtle bugs. Since there seems to be no special benefit from calling listener under synchronized, Move the call outside. Of course, you still want to read `memoryRevokingRequested` inside lock.